client/src/main/java/org/asynchttpclient/cookie/ThreadSafeCookieStore.java

```diff
@@ -145,45 +149,64 @@ private void add(String requestDomain, String requestPath, Cookie cookie) {
     String keyDomain = pair.getKey();
     boolean hostOnly = pair.getValue();
     String keyPath = cookiePath(cookie.path(), requestPath);
-    CookieKey key = new CookieKey(cookie.name().toLowerCase(), keyDomain, keyPath);
+    CookieKey key = new CookieKey(cookie.name().toLowerCase(), keyPath);
 
     if (hasCookieExpired(cookie, 0))
-      cookieJar.remove(key);
-    else
-      cookieJar.put(key, new StoredCookie(cookie, hostOnly, cookie.maxAge() != Cookie.UNDEFINED_MAX_AGE));
+      cookieJar.getOrDefault(keyDomain, Collections.emptyMap()).remove(key);
+    else {
+      cookieJar.putIfAbsent(keyDomain, new ConcurrentHashMap<>());
+      cookieJar.get(keyDomain).put(key, new StoredCookie(cookie, hostOnly, cookie.maxAge() != Cookie.UNDEFINED_MAX_AGE));
+    }
   }
 
   private List<Cookie> get(String domain, String path, boolean secure) {
 
     final boolean[] removeExpired = {false};
+    final List<Cookie> results = new ArrayList<>();
```

Comment:
Please reduce allocations here.
Default ArrayList size is 10, it's pretty unlikely that we get more than 2 cookies for a given domain. Also, it can be allocated lazily on first match (lots of AHC usage won't involve cookies at all).